includes/validation/class-amp-validation-manager.php

```diff
@@ -1893,38 +1890,105 @@ public static function validate_url( $url ) {
 		);
 	}
 
+	/**
+	 * Serialize validation error messages.
+	 *
+	 * In order to safely pass validation error messages through redirects with query parameters, they must be serialized
+	 * with a HMAC for security. The messages contain markup so the HMAC prevents tampering.
+	 *
+	 * @since 1.4.2
+	 * @see AMP_Validation_Manager::unserialize_validation_error_messages()
+	 *
+	 * @param string[] $messages Messages.
+	 * @return string Serialized.
+	 */
+	public static function serialize_validation_error_messages( $messages ) {
+		$encoded_messages = base64_encode( wp_json_encode( array_unique( $messages ) ) ); // phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.obfuscation_base64_encode
+		return wp_hash( $encoded_messages ) . ':' . $encoded_messages;
```

Comment:
Just want to note that the `wp_hash()` function uses `md5`, which is usually to be avoided for security-related logic.

Using the PHP `hash_hmac()` directly would allow the use of a stronger algo.

The context does not seem too critical to me right now, but I am not a security expert. (IANASE?)

Reply:
TIL: IANASE

Note that `\WP_Customize_Widgets::get_instance_hash_key()` uses `wp_hash()` as opposed to `hash_hmac()`, so I think it's fine. And you're right, it's not critical here, as the strings being output are being passed through `wp_kses()` in any case.